tests/ibc/ibc_middleware_test.go

```diff
@@ -0,0 +1,814 @@
+	suite.Require().True(escrowedBal.IsZero(), "escrowed balance should be un-escrowed after receiving the packet")
+	balAfterUnescrow := evmApp.Erc20Keeper.BalanceOf(evmCtx, nativeErc20.ContractAbi, nativeErc20.ContractAddr, senderEthAddr)
+	suite.Require().Equal(nativeErc20.InitialBal.String(), balAfterUnescrow.String())
+}
+
+func (suite *MiddlewareTestSuite) TestOnAcknowledgementPacket() {
+	var (
+		packet channeltypes.Packet
+		ack    []byte
+	)
+
+	testCases := []struct {
+		name           string
+		malleate       func()
+		onSendRequired bool
+		expError       string
+	}{
+		{
+			name:           "pass",
+			malleate:       nil,
+			onSendRequired: false,
+			expError:       "",
+		},
+		{
+			name: "pass: refund escrowed token",
+			malleate: func() {
+				ackErr := channeltypes.NewErrorAcknowledgement(errors.New("error"))
+				ack = ackErr.Acknowledgement()
+			},
+			onSendRequired: true,
+			expError:       "",
+		},
+		{
+			name: "fail: malformed packet data",
+			malleate: func() {
+				packet.Data = []byte("malformed data")
+			},
+			onSendRequired: false,
+			expError:       "cannot unmarshal ICS-20 transfer packet data",
+		},
+		{
+			name: "fail: empty ack",
+			malleate: func() {
+				ack = []byte{}
+			},
+			onSendRequired: false,
+			expError:       "cannot unmarshal ICS-20 transfer packet acknowledgement",
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			ctxA := suite.evmChainA.GetContext()
+			evmApp := suite.evmChainA.App.(*evmd.EVMD)
+
+			bondDenom, err := evmApp.StakingKeeper.BondDenom(ctxA)
+			suite.Require().NoError(err)
+
+			sendAmt := ibctesting.DefaultCoinAmount
+			sender := suite.evmChainA.SenderAccount.GetAddress()
+			receiver := suite.chainB.SenderAccount.GetAddress()
+
+			packetData := transfertypes.NewFungibleTokenPacketData(
+				bondDenom,
+				sendAmt.String(),
+				sender.String(),
+				receiver.String(),
+				"",
+			)
+
+			path := suite.pathAToB
+			packet = channeltypes.Packet{
+				Sequence:           1,
+				SourcePort:         path.EndpointA.ChannelConfig.PortID,
+				SourceChannel:      path.EndpointA.ChannelID,
+				DestinationPort:    path.EndpointB.ChannelConfig.PortID,
+				DestinationChannel: path.EndpointB.ChannelID,
+				Data:               packetData.GetBytes(),
+				TimeoutHeight:      suite.chainB.GetTimeoutHeight(),
+				TimeoutTimestamp:   0,
+			}
+
+			ack = channeltypes.NewResultAcknowledgement([]byte{1}).Acknowledgement()
+			if tc.malleate != nil {
+				tc.malleate()
+			}
+
+			transferStack, ok := evmApp.GetIBCKeeper().PortKeeper.Route(transfertypes.ModuleName)
+			suite.Require().True(ok)
+
+			sourceChan := suite.pathAToB.EndpointA.GetChannel()
+			onAck := func() error {
+				return transferStack.OnAcknowledgementPacket(
+					ctxA,
+					sourceChan.Version,
+					packet,
+					ack,
+					receiver,
+				)
+			}
+			if tc.onSendRequired {
```

Comment:
Do an escrow check for this

Reply:
I just added the checking logic, thanks

https://github.com/cosmos/evm/blob/7e1f25917fe16e985a3cda51622474feee3d44dc/tests/ibc/ibc_middleware_test.go#L406-L435